File: android/src/main/kotlin/com/snowplowanalytics/snowplow_flutter_tracker/readers/events/SelfDescribingJsonReader.kt
// Copyright (c) 2022 Snowplow Analytics Ltd. All rights reserved.
//
// This program is licensed to you under the Apache License Version 2.0,
// and you may not use this file except in compliance with the Apache License Version 2.0.
// You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the Apache License Version 2.0 is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.

package com.snowplowanalytics.snowplow_flutter_tracker.readers.events

import com.snowplowanalytics.snowplow.event.SelfDescribing
import com.snowplowanalytics.snowplow.payload.SelfDescribingJson

class SelfDescribingJsonReader(val values: Map<String, Any>) {
    val schema: String by values
    val data: Map<String, Any> by values

    fun toSelfDescribingJson(): SelfDescribingJson {
        return SelfDescribingJson(schema, data)
    }

    fun toSelfDescribing(): SelfDescribing {
        return SelfDescribing(toSelfDescribingJson())
    }
}

File: ios/Classes/readers/events/SelfDescribingJsonReader.swift
//
//  SelfDescribingJsonReader.swift
//  snowplow_flutter_tracker
//
//  Created by Matus Tomlein on 04/01/2022.
//

import Foundation
import SnowplowTracker

struct SelfDescribingJsonReader: Decodable {
    let schema: String
}

extension SelfDescribingJsonReader {
    func toSelfDescribingJson(arguments: [String: Any]) -> SelfDescribingJson? {
        if let data = arguments["data"] as? [String: Any] {
            return SelfDescribingJson(schema: schema, andDictionary: data)
        }
        return nil
    }
    
    func toSelfDescribing(arguments: [String: Any]) -> SelfDescribing? {
        if let json = toSelfDescribingJson(arguments: arguments) {
            return SelfDescribing(eventData: json)
        }
        return nil
    }
}
